### ISSUE TITLE
 pullRemoteCache  -> remote cache not found for 0.0.0-761fa7174f

### ISSUE THREAD

**Issue by tyrcho** [sbt/sbt#5725]
## steps

I updated sbt to 1.4.0-M1 and added to my build.sbt this line

    ThisBuild / pushRemoteCacheTo := Some(MavenCache("local-cache", file("/tmp/remote-cache")))

Complete project (very basic, and I had the same issues on 3 different projects I tried)

https://gitlab.com/tyrcho/approvals-scala/-/merge_requests/40/diffs

From sbt shell  :

```
[IJ]sbt:approvals-scala> pushRemoteCache
[info] Compiling 1 Scala source to /Users/michel/git/katas/approvals-scala/target/scala-2.12/test-classes ...
[info] Wrote /Users/michel/git/katas/approvals-scala/target/scala-2.12/approvals-scala_2.12-0.1.0-SNAPSHOT.pom
[info] Done compiling.
[info] 	published approvals-scala_2.12 to file:/tmp/remote-cache/com/github/writethemfirst/approvals-scala_2.12/0.0.0-761fa71/approvals-scala_2.12-0.0.0-761fa71.pom
[info] 	published approvals-scala_2.12 to file:/tmp/remote-cache/com/github/writethemfirst/approvals-scala_2.12/0.0.0-761fa71/approvals-scala_2.12-0.0.0-761fa71-cached-compile.jar
[info] 	published approvals-scala_2.12 to file:/tmp/remote-cache/com/github/writethemfirst/approvals-scala_2.12/0.0.0-761fa71/approvals-scala_2.12-0.0.0-761fa71-cached-test.jar
[success] Total time: 1 s, completed Aug 5, 2020, 3:19:31 PM
[IJ]sbt:approvals-scala> pullRemoteCache
[info] remote cache not found for 0.0.0-761fa7174f
[info] remote cache not found for 0.0.0-e3e756ff7a
[info] remote cache not found for 0.0.0-1b3a414692
[info] remote cache not found for 0.0.0-fddd0ac62c
[info] remote cache not found for 0.0.0-be00317f0f
[success] Total time: 0 s, completed Aug 5, 2020, 3:19:35 PM
[IJ]sbt:approvals-scala> 
```

## problem

the cache is not found and cannot be reused.

## expectation

I expect `pullRemoteCache` to find the cache.

## notes

The cache is actually written in the folder as expected:
```
/tmp/remote-cache/com/github/writethemfirst/approvals-scala_2.12/0.0.0-761fa71:
approvals-scala_2.12-0.0.0-761fa71-cached-compile.jar      approvals-scala_2.12-0.0.0-761fa71-cached-test.jar         approvals-scala_2.12-0.0.0-761fa71.pom
approvals-scala_2.12-0.0.0-761fa71-cached-compile.jar.md5  approvals-scala_2.12-0.0.0-761fa71-cached-test.jar.md5     approvals-scala_2.12-0.0.0-761fa71.pom.md5
approvals-scala_2.12-0.0.0-761fa71-cached-compile.jar.sha1 approvals-scala_2.12-0.0.0-761fa71-cached-test.jar.sha1    approvals-scala_2.12-0.0.0-761fa71.pom.sha1
```

I tried with and without a clean step between push and pull and it changes nothing.

I think the issue is with the id.

Push used 761fa71 and pull is looking for 761fa7174f

@eed3si9n 

I'm very excited with this cache feature (since hoarder stopped working months ago) for a dozen of projects, both personal and with my team at work. If I can help narrowing down the issue, reproducing on different context I'd be happy to help. I can't fix it, sbt codebase is out of my comfort zone though :D

**Comment by eed3si9n** [comment]
It works for me on my machine. What do you see if you typed in `debug` then `pullRemoteCache`

```
akka > debug
akka > akka-actor/pullRemoteCache
[debug] Attempting to fetch com.typesafe.akka % akka-actor_2.12 % 0.0.0-394b4fba9c, com.typesafe.akka % akka-actor_2.12 % 0.0.0-394b4fba9c. This operation may fail.
[info] remote cache not found for 0.0.0-394b4fba9c
[debug] Attempting to fetch com.typesafe.akka % akka-actor_2.12 % 0.0.0-fe3fcae25c, com.typesafe.akka % akka-actor_2.12 % 0.0.0-fe3fcae25c. This operation may fail.
[debug] Files retrieved for com.typesafe.akka % akka-actor_2.12 % 0.0.0-fe3fcae25c, com.typesafe.akka % akka-actor_2.12 % 0.0.0-fe3fcae25c:
[debug] /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-fe3fcae25c/akka-actor_2.12-0.0.0-fe3fcae25c-cached-compile.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-fe3fcae25c/akka-actor_2.12-0.0.0-fe3fcae25c-cached-test.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-fe3fcae25c/akka-actor_2.12-0.0.0-fe3fcae25c-cached-compile.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-fe3fcae25c/akka-actor_2.12-0.0.0-fe3fcae25c-cached-test.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-fe3fcae25c/akka-actor_2.12-0.0.0-fe3fcae25c-cached-compile.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-fe3fcae25c/akka-actor_2.12-0.0.0-fe3fcae25c-cached-test.jar
[success] Total time: 1 s, completed Aug 11, 2020 11:21:35 PM
```

**Comment by tyrcho** [comment]
Thanks for having a look !

Here is what I got:

```
[IJ]sbt:approvals-scala> approvals-scala/pullRemoteCache
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-7016caf695, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-7016caf695. This operation may fail.
[info] remote cache not found for 0.0.0-7016caf695
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-761fa7174f, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-761fa7174f. This operation may fail.
[info] remote cache not found for 0.0.0-761fa7174f
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-e3e756ff7a, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-e3e756ff7a. This operation may fail.
[info] remote cache not found for 0.0.0-e3e756ff7a
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-1b3a414692, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-1b3a414692. This operation may fail.
[info] remote cache not found for 0.0.0-1b3a414692
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-fddd0ac62c, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-fddd0ac62c. This operation may fail.
[info] remote cache not found for 0.0.0-fddd0ac62c
[success] Total time: 0 s, completed Aug 12, 2020, 10:18:08 AM
[IJ]sbt:approvals-scala> 
```

**Comment by tyrcho** [comment]
I don't think the issue is related to my machine, we can reproduce it on a docker container:

```
docker run -ti hseeberger/scala-sbt:11.0.6_1.3.9_2.12.10 bash                  

# inside docker
git clone -b remote_cache https://gitlab.com/tyrcho/approvals-scala
cd approvals-scala/
sbt
# [info] [launcher] getting org.scala-sbt sbt 1.4.0-M1  (this may take some time)...
# ...
# [info] welcome to sbt 1.4.0-M1 (Oracle Corporation Java 11.0.6)
# more download ...

# inside sbt
compile
debug
pushRemoteCache
pullRemoteCache

[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-7016caf695, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-7016caf695. This operation may fail.
[info] remote cache not found for 0.0.0-7016caf695
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-761fa7174f, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-761fa7174f. This operation may fail.
[info] remote cache not found for 0.0.0-761fa7174f
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-e3e756ff7a, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-e3e756ff7a. This operation may fail.
[info] remote cache not found for 0.0.0-e3e756ff7a
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-1b3a414692, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-1b3a414692. This operation may fail.
[info] remote cache not found for 0.0.0-1b3a414692
[debug] Attempting to fetch com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-fddd0ac62c, com.github.writethemfirst % approvals-scala_2.12 % 0.0.0-fddd0ac62c. This operation may fail.
[info] remote cache not found for 0.0.0-fddd0ac62c
[success] Total time: 0 s, completed Aug 12, 2020, 8:30:29 AM
```

yield the same result

**Comment by tyrcho** [comment]
I tried to rename these files `/tmp/remote-cache/com/github/writethemfirst/approvals-scala_2.12/0.0.0-7016caf/approvals-scala_2.12-0.0.0-7016caf-*`

to `/tmp/remote-cache/com/github/writethemfirst/approvals-scala_2.12/0.0.0-7016caf695/approvals-scala_2.12-0.0.0-7016caf695-*`

and the command `pullRemoteCache` worked.

So it seems the issue is with `pushRemoteCache` using a id too short (7 chars instead of 10 expected by pull)

**Comment by tyrcho** [comment]
I tried on another project https://gitlab.com/colisweb-open-source/scala/skills and I have the same issue, again the id used in push is shorter

```
sbt:skills> pushRemoteCache
[info] Wrote /root/skills/target/scala-2.13/skills_2.13-0.1.0-SNAPSHOT.pom
[info] 	published skills_2.13 to file:/tmp/remote-cache/com/colisweb/skills_2.13/0.0.0-3c7aea1/skills_2.13-0.0.0-3c7aea1.pom

sbt:skills> pullRemoteCache
[info] remote cache not found for 0.0.0-3c7aea1982

```

**Comment by tyrcho** [comment]
I tried to reproduce the issue with akka, and I couldn't.

```
sbt:akka-actor> set every targetSystemJdk := true
sbt:akka-actor> compile
sbt:akka-actor> pushRemoteCache
[info] Wrote /root/akka/akka-actor/target/scala-2.12/akka-actor_2.12-2.6.8+35-4826477c+20200812-1012.pom
[info] Validating all packages are set private or exported for OSGi explicitly...
[warn] bnd: Unused Private-Package instructions, no such package(s) on the class path: [akka.osgi.impl]
[info] 	published akka-actor_2.12 to file:/tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36.pom
[info] 	published akka-actor_2.12 to file:/tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-compile.jar
[info] 	published akka-actor_2.12 to file:/tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-test.jar
[success] Total time: 3 s, completed Aug 12, 2020, 10:13:51 AM


sbt:akka-actor> pullRemoteCache
[debug] Attempting to fetch com.typesafe.akka % akka-actor_2.12 % 0.0.0-4826477c36, com.typesafe.akka % akka-actor_2.12 % 0.0.0-4826477c36. This operation may fail.
[debug] Files retrieved for com.typesafe.akka % akka-actor_2.12 % 0.0.0-4826477c36, com.typesafe.akka % akka-actor_2.12 % 0.0.0-4826477c36:
[debug] /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-compile.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-test.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-compile.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-test.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-compile.jar, /tmp/remote-cache/com/typesafe/akka/akka-actor_2.12/0.0.0-4826477c36/akka-actor_2.12-0.0.0-4826477c36-cached-test.jar
[success] Total time: 0 s, completed Aug 12, 2020, 10:15:51 AM
```

So this seems to be an issue related to some specificity of the sbt project.

**Comment by tyrcho** [comment]
I tried on the simplest scala project I could find : https://github.com/tyrcho/scala-maven-template

No plugins. 3 lines build.sbt.

And I have the issue:
```
[info]  published scala-template_2.13 to file:/tmp/remote-cache/scala-template/scala-template_2.13/0.0.0-1605913/scala-template_2.13-0.0.0-1605913-cached-test.jar
sbt:scala template> pullRemoteCache
[info] remote cache not found for 0.0.0-1605913ecf
```

So my understanding is that something must be missing in the simple projects I'm testing on.

I don't think of anymore investigation I can do on my side @eed3si9n

**Comment by tyrcho** [comment]
if I "hardcode" the cacheId in my build.sbt with `ThisBuild / remoteCacheId := "1605913ecf"` then it writes in the correct folder and pullRemoteCache succeeds.

So the issue can be narrowed down to *Why does the pushRemoteCache gets an id with 7 characters instead of 10 ?*

**Comment by tyrcho** [comment]
Adding this 
`ThisBuild / remoteCacheId := scala.sys.process.Process("git rev-parse HEAD").!!.trim.take(10)`

On top of
`ThisBuild / pushRemoteCacheTo := Some(MavenCache("local-cache", file("/tmp/remote-cache")))`

Seems to be workaround.

By looking at the history of `main/src/main/scala/sbt/internal/RemoteCache.scala` I have some hope that this issue was fixed after sbt 1.4.0-M1 was published .

**Comment by eed3si9n** [comment]
Yes. I've forgotten about https://github.com/sbt/sbt/issues/5621. It will be fixed in the next pre-release.
